### PR TITLE
No .qm at home, will it work there?

### PR DESCRIPTION
**Pull request by nirvn**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -260,3 +260,23 @@ if(WITH_VCPKG)
 
   add_dependencies(qfield deploy)
 endif()
+
+# lrelease .ts files
+find_package(Qt5 COMPONENTS LinguistTools)
+
+file(GLOB TS_FILES ${CMAKE_SOURCE_DIR}/i18n/*.ts)
+set_source_files_properties(${TS_FILES} PROPERTIES OUTPUT_LOCATION "${CMAKE_SOURCE_DIR}/i18n")
```
[the remark this answers is not in the document]
Comment (3nids) [reply CMakeLists.txt:268]: but shouldn't you remove the current QRC file too?
here https://github.com/opengisch/QField/blob/baae03ed2383d407b66489fc755fe5583889408a/src/app/CMakeLists.txt#L19

I guess this is the one causing issue as the error is in build/src/app (while your changes are on top directory)

[the remark this answers is not in the document]
Comment (nirvn) [reply CMakeLists.txt]: @3nids , moved this into src/app/CMakeLists.txt and it works :partying_face: 

The only problem is I can't get it to work other than by having the .qm files generated in the source tree, which isn't ideal.

That said, it's all dynamic, woupidou.

File: src/app/CMakeLists.txt [hunk 2]
```diff
@@ -10,13 +10,33 @@ set(QML_IMPORT_PATH "${CMAKE_SOURCE_DIR}/src/qml/imports")
 file(GLOB_RECURSE QML_SOURCES "${CMAKE_SOURCE_DIR}/src/qml/*.qml")
 add_custom_target(qfield_qml SOURCES ${QML_SOURCES})
 
+# lrelease .ts files
+find_package(Qt5 COMPONENTS LinguistTools)
+
+file(GLOB TS_FILES ${CMAKE_SOURCE_DIR}/i18n/*.ts)
+set_source_files_properties(${TS_FILES} PROPERTIES OUTPUT_LOCATION "${CMAKE_SOURCE_DIR}/i18n")
```
Comment (3nids) [inline src/app/CMakeLists.txt:17]:
```suggestion
set_source_files_properties(${TS_FILES} PROPERTIES OUTPUT_LOCATION "${CMAKE_BINARY_DIR}/i18n")
```
if you want to have them in build directory

Comment (nirvn) [reply src/app/CMakeLists.txt:17]: @3nids , I've tried that, without success so far.

**Comment by 3nids** [comment]
also, don't forget to remove the original qrc file in i18n

File: src/app/CMakeLists.txt (hunk 2 above)
Comment (3nids) [reply src/app/CMakeLists.txt:17]: from the docs 

> By default, the qm files will be placed in the root level of the build directory. To change this, you can set OUTPUT_LOCATION as a property of the source .ts file.

if you only write `set_source_files_properties(${TS_FILES} PROPERTIES OUTPUT_LOCATION "i18n")`, what happens?

Comment (nirvn) [reply src/app/CMakeLists.txt:17]: folder gets created but no qm inside


ninja: error: 'i18n/qfield_sv.qm', needed by 'src/app/qfield_autogen/P4RBTXXXTD/qrc_translations.cpp', missing and no known rule to make it
20:16:08: The process "/usr/bin/cmake" exited with code 1.
Error while building/deploying project QField (kit: Desktop)

Comment (3nids) [reply src/app/CMakeLists.txt:17]: and nothing anywhere?

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for ad776692f5665ccced25a8495fed2fbe35d83aea:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-4338fe310be5dde0e74e60136b857039ebd37dad-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-4338fe310be5dde0e74e60136b857039ebd37dad-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-4338fe310be5dde0e74e60136b857039ebd37dad-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-4338fe310be5dde0e74e60136b857039ebd37dad-x86.apk)

File: src/app/CMakeLists.txt (hunk 2 above)
Comment (3nids) [reply src/app/CMakeLists.txt:17]: you can maybe copy the Qt macro https://github.com/qt/qttools/blob/5.15.2/src/linguist/Qt5LinguistToolsMacros.cmake#L112
and output some debug messages to see what's going on

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 3cd63375ef9ba3578cde7dae7a764189f2904984:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-9e091e8b8d853cb5213a63d18c13040213db1a98-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-9e091e8b8d853cb5213a63d18c13040213db1a98-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-9e091e8b8d853cb5213a63d18c13040213db1a98-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-9e091e8b8d853cb5213a63d18c13040213db1a98-x86.apk)

**Comment by nirvn** [comment]
Yay, working, APKs have all the translations on transifex auto generated!

File: scripts/ci/pull_translations.sh [hunk 3]
```diff
@@ -13,7 +13,3 @@ echo "::endgroup::"
 echo "::group::android specific translations"
 for x in android/res/values-*_*;do mv $x $(echo $x | sed -e 's/_/-r/') ;done
 echo "::endgroup::"
-
-echo "::group::lrelease"
-find ${DIR}/../../i18n -type f -name "*.ts" | xargs lrelease
```
Comment (3nids) [inline scripts/ci/pull_translations.sh:18]: 😢

File: src/app/CMakeLists.txt [hunk 4]
```diff
@@ -10,13 +10,33 @@ set(QML_IMPORT_PATH "${CMAKE_SOURCE_DIR}/src/qml/imports")
...
+file(APPEND ${TRANSLATIONS_QRC} "\n    </qresource>\n</RCC>")
+
 if (ANDROID)
   add_library(qfield SHARED
     main.cpp
     ${RESOURCES}
     ${CMAKE_SOURCE_DIR}/images/images.qrc
     ${CMAKE_SOURCE_DIR}/resources/resources.qrc
-    ${CMAKE_SOURCE_DIR}/i18n/translations.qrc
+    ${CMAKE_BINARY_DIR}/translations.qrc
```
Comment (3nids) [inline src/app/CMakeLists.txt:39]:
```suggestion
    ${TRANSLATIONS_QRC}
```

File: src/app/CMakeLists.txt [hunk 5]
```diff
@@ -27,7 +47,7 @@ else()
...
```
Comment (3nids) [inline src/app/CMakeLists.txt:50]:
```suggestion
    ${TRANSLATIONS_QRC}
```

**Review by 3nids — APPROVED** [review]

File: scripts/ci/pull_translations.sh (hunk 3 above)
Comment (m-kuhn) [reply scripts/ci/pull_translations.sh:18]: Time to learn to let go

Comment (3nids) [reply scripts/ci/pull_translations.sh:18]: tough times

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for b4ca80f4ed855b839f21d269076e1fb8b5ef300d:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-eaec73dbfa2c2f7f5d3b0bf8a1412d5e55f21949-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-eaec73dbfa2c2f7f5d3b0bf8a1412d5e55f21949-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-eaec73dbfa2c2f7f5d3b0bf8a1412d5e55f21949-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-eaec73dbfa2c2f7f5d3b0bf8a1412d5e55f21949-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for e20fdd86398e24ff8d7cf6f392b8e0063a618327:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-001635aa79fac31fad7d2e5c49e4636bccdabbf2-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-001635aa79fac31fad7d2e5c49e4636bccdabbf2-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-001635aa79fac31fad7d2e5c49e4636bccdabbf2-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2190-001635aa79fac31fad7d2e5c49e4636bccdabbf2-x86.apk)

**Comment by nirvn** [comment]
Mission accomplished:
- the list of available languages for users to pick from is now dynamically generated (a cool little trick iterating through the qrc embedded .qm  files)
- the .ts files are now removed from the source tree
- the pull_translation.sh script has been updated to only pull translations that are >=50% done
- vcpkg packages now embed translations again

**Comment by m-kuhn** [comment]
Wow

**Comment by nirvn** [comment]
@m-kuhn , I just checked the vcpkg package during a brief booting on windows. Unfortunately the qm files aren't generated / added there, which is a bit odd. We're not getting a cmake error, which I assume means the qt linguist tools development environment is there (otherwise we'd get a macro not found issue using qt5_add_translation), yet no .qm is being generated.

Shot in the dark: the .ts files pulled via tx pull aren't landing in the right directory?

**Pull request merged**